keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -443,15 +459,45 @@ class ApplicationService(
     }
   }
 
+  private fun getArtifactVersionByPromotionStatus(
+    environmentName: String,
+    promotionStatus: PromotionStatus,
+    version: String?,
+    artifactInfoInEnvironment: Map<String, List<StatusInfoForArtifactInEnvironment>>,
+    allVersions: List<PublishedArtifact>
```

Comment:
Consider standardizing on a name for this object. You have `allVersions` here and `versions` above. I prefer `allVersions`.